src/charm.py

```diff
@@ -694,39 +697,148 @@ def _activate_charm(self) -> None:
             self._start_service()
             self.model.unit.status = ActiveStatus()
 
-    def _on_add_admin_user_action(self, event: ActionEvent) -> None:
-        """Add a new admin user to Discourse.
+    def _on_promote_user_action(self, event: ActionEvent) -> None:
+        """Promote a user to a specific trust level.
 
         Args:
-            event: Event triggering the add_admin_user action.
+            event: Event triggering the promote_user action.
         """
-        email = event.params["email"]
-        password = event.params["password"]
         container = self.unit.get_container(CONTAINER_NAME)
-        if container.can_connect():
-            process = container.exec(
-                [
-                    os.path.join(DISCOURSE_PATH, "bin/bundle"),
-                    "exec",
-                    "rake",
-                    "admin:create",
-                ],
-                stdin=f"{email}\n{password}\n{password}\nY\n",
-                working_dir=DISCOURSE_PATH,
-                user=CONTAINER_APP_USERNAME,
-                environment=self._create_discourse_environment_settings(),
-                timeout=60,
+        if not container.can_connect():
+            event.fail("Unable to connect to container, container is not ready")
+            return
+
+        email = event.params["email"]
+
+        user_exists = container.exec(
+            [os.path.join(DISCOURSE_PATH, "bin/bundle"), "exec", "rake", f"users:exists[{email}]"],
+            working_dir=DISCOURSE_PATH,
+            user=CONTAINER_APP_USERNAME,
+            environment=self._create_discourse_environment_settings(),
+        )
+        try:
+            user_exists.wait_output()
+        except ExecError as ex:
+            if ex.exit_code == 2:
+                event.fail(f"User with email {email} does not exist")
+            else:
+                event.fail(
+                    f"Error checking if user with email {email} exists: {ex.stdout}"  # type: ignore
+                )
+            return
+
+        process = container.exec(
+            [
+                os.path.join(DISCOURSE_PATH, "bin/bundle"),
+                "exec",
+                "rake",
+                "admin:create",
+            ],
+            stdin=f"{email}\nn\nY\n",
+            working_dir=DISCOURSE_PATH,
+            user=CONTAINER_APP_USERNAME,
+            environment=self._create_discourse_environment_settings(),
+            timeout=60,
+        )
+        try:
+            process.wait_output()
+            event.set_results({"user": email})
+        except ExecError as ex:
+            event.fail(
+                f"Failed to make user with email {email} an admin: {ex.stdout}"  # type: ignore
             )
-            try:
-                process.wait_output()
-                event.set_results({"user": f"{email}"})
-            except ExecError as ex:
+
+    def _on_create_user_action(self, event: ActionEvent) -> None:
```

Comment:
The logic of this function is simple, but the function is very long and so it is difficult to see the real logic. I believe splitting it into several functions will help with readability. This will imply moving the code that calls pebble to its own functions.

Reply:
I have split the users exist call into a separate method, since this is reusable. I have tried to tidy up the methods, but I don't think moving the code that calls pebble will help much -

it will just create a number of non-reusable methods that greater obfuscate the action behaviour. Also, this would not be consistent with the anonymize user action. 

With the user exists code being moved, the promote action is now roughly the same size as that anonymize action and just as readable in my opinion, and the create action is really only more complex due to the user activation at the end

Reply:
Separating the `_user_exists` is an improvement! Feel free to close this comment.

My opinion, in general, is that it is not only about being able to reuse functions, but about readability (and when it gets complex, separating by abstractions levels). For example, for create_user, a function like this could separate it from the pebble calls (and the pebble calls are hidden in _user_exists, _create_user and _activate_user):
```
        container = self.unit.get_container(CONTAINER_NAME)
        if not container.can_connect():
            event.fail("Unable to connect to container, container is not ready")
            return

        email = event.params["email"]
        password = self._generate_password(16)
        admin = bool(event.params.get("admin", False))
        active = bool(event.params.get("active", False))

        if self._user_exists(email):
            event.fail(f"User with email {email} already exists")
            return

        self._create_user(email, password, admin)

        if not admin or active:
            self._activate_user(email)

        event.set_results({"user": email, "password": password})
```